File: plugins/tunnel-mcp/scripts/tunnel_mcp_cli/state.py
from __future__ import annotations

import dataclasses
import datetime as dt
import json
import os
import re
import shlex
from pathlib import Path
from typing import Any

_ALIAS_CHARS = re.compile(r"[^a-z0-9-]+")
_SECRET_PATTERNS = [
    re.compile(r"sk-[A-Za-z0-9_-]{12,}"),
    re.compile(r"(?i)\bBearer\s+[A-Za-z0-9._~+/=-]{12,}"),
    re.compile(r"(?i)\bAuthorization\s*:"),
    re.compile(
        r"(?i)(api[_-]?key|access[_-]?token|refresh[_-]?token|id[_-]?token|password|secret)=\S+"
    ),
    re.compile(r"(?i)\b(OPENAI_ADMIN_KEY|OPENAI_API_KEY|CONTROL_PLANE_API_KEY)="),
    re.compile(r"://[^/\s:@]+:[^/\s:@]+@"),
]
_SECRET_FLAG_PATTERN = re.compile(
    r"(?i)^-{1,2}[a-z0-9_-]*(api[_-]?key|access[_-]?token|refresh[_-]?token|id[_-]?token|password|secret)[a-z0-9_-]*$"
)
_SECRET_REFERENCE_PATTERN = re.compile(
    r"^(env:[A-Za-z_][A-Za-z0-9_]*|file:.+|\$[A-Za-z_][A-Za-z0-9_]*|\$\{[A-Za-z_][A-Za-z0-9_]*\})$"
)


class StateError(RuntimeError):
    pass


@dataclasses.dataclass(frozen=True)
class AliasRecord:
    alias: str
    tunnel_id: str
    name: str
    admin_profile: str = ""
    description: str = ""
    organization_ids: tuple[str, ...] = ()
    workspace_ids: tuple[str, ...] = ()
    tenant_ids: tuple[str, ...] = ()
    config_path: str = ""
    health_url_file: str = ""
    updated_at: str = ""

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> "AliasRecord":
        return cls(
            alias=str(raw.get("alias", "")),
            tunnel_id=str(raw.get("tunnel_id", "")),
            name=str(raw.get("name", "")),
            admin_profile=str(raw.get("admin_profile", "")),
            description=str(raw.get("description", "")),
            organization_ids=tuple(str(v) for v in raw.get("organization_ids", [])),
            workspace_ids=tuple(str(v) for v in raw.get("workspace_ids", [])),
            tenant_ids=tuple(str(v) for v in raw.get("tenant_ids", [])),
            config_path=str(raw.get("config_path", "")),
            health_url_file=str(raw.get("health_url_file", "")),
            updated_at=str(raw.get("updated_at", "")),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "alias": self.alias,
            "tunnel_id": self.tunnel_id,
            "name": self.name,
            "admin_profile": self.admin_profile,
            "description": self.description,
            "organization_ids": list(self.organization_ids),
            "workspace_ids": list(self.workspace_ids),
            "tenant_ids": list(self.tenant_ids),
            "config_path": self.config_path,
            "health_url_file": self.health_url_file,
            "updated_at": self.updated_at,
        }


@dataclasses.dataclass(frozen=True)
class ProcessRecord:
    alias: str
    tunnel_id: str
    config_path: str
    health_url_file: str
    target_kind: str
    target_value: str
    command: str
    started_at: str
    admin_profile: str = ""
    mode: str = "tmux"
    session_name: str = ""
    pid: int = 0
    log_path: str = ""

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> "ProcessRecord":
        return cls(
            alias=str(raw.get("alias", "")),
            tunnel_id=str(raw.get("tunnel_id", "")),
            admin_profile=str(raw.get("admin_profile", "")),
            mode=str(raw.get("mode", "tmux")),
            session_name=str(raw.get("session_name", "")),
            pid=int(raw.get("pid", 0) or 0),
            config_path=str(raw.get("config_path", "")),
            health_url_file=str(raw.get("health_url_file", "")),
            target_kind=str(raw.get("target_kind", "")),
            target_value=str(raw.get("target_value", "")),
            command=str(raw.get("command", "")),
            log_path=str(raw.get("log_path", "")),
            started_at=str(raw.get("started_at", "")),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "alias": self.alias,
            "tunnel_id": self.tunnel_id,
            "admin_profile": self.admin_profile,
            "mode": self.mode,
            "session_name": self.session_name,
            "pid": self.pid,
            "config_path": self.config_path,
            "health_url_file": self.health_url_file,
            "target_kind": self.target_kind,
            "target_value": self.target_value,
            "command": self.command,
            "log_path": self.log_path,
            "started_at": self.started_at,
        }


@dataclasses.dataclass(frozen=True)
class AdminProfile:
    name: str
    control_plane_base_url: str
    admin_key: str
    updated_at: str = ""

    @classmethod
    def from_dict(cls, raw: dict[str, Any]) -> "AdminProfile":
        return cls(
            name=str(raw.get("name", "")),
            control_plane_base_url=str(raw.get("control_plane_base_url", "")),
            admin_key=str(raw.get("admin_key", "")),
            updated_at=str(raw.get("updated_at", "")),
        )

    def to_dict(self) -> dict[str, Any]:
        return {
            "name": self.name,
            "control_plane_base_url": self.control_plane_base_url,
            "admin_key": self.admin_key,
            "updated_at": self.updated_at,
        }


def utc_now() -> str:
    return dt.datetime.now(dt.UTC).replace(microsecond=0).isoformat().replace("+00:00", "Z")


def normalize_alias(alias: str) -> str:
    lowered = alias.strip().lower()
    lowered = lowered.replace("_", "-")
    lowered = _ALIAS_CHARS.sub("-", lowered)
    lowered = re.sub(r"-{2,}", "-", lowered).strip("-")
    if not lowered:
        raise StateError("alias must contain at least one ASCII letter or number")
    return lowered


def state_root() -> Path:
    codex_home = os.environ.get("CODEX_HOME")
    if codex_home:
        return Path(codex_home).expanduser() / "tunnel-mcp"
    return Path.home() / ".codex" / "tunnel-mcp"


def ensure_state_dirs(root: Path | None = None) -> Path:
    root = root or state_root()
    for subdir in (root, root / "configs", root / "health", root / "logs"):
        subdir.mkdir(parents=True, exist_ok=True)
    return root


def admin_profiles_path(root: Path | None = None) -> Path:
    return _path(root, "admin_profiles.yaml")


def load_aliases(root: Path | None = None) -> dict[str, AliasRecord]:
    raw = _load_json_map(_path(root, "aliases.yaml"))
    return {alias: AliasRecord.from_dict(value) for alias, value in raw.items()}


def save_aliases(records: dict[str, AliasRecord], root: Path | None = None) -> None:
    _save_json_map(_path(root, "aliases.yaml"), {k: records[k].to_dict() for k in sorted(records)})


def load_processes(root: Path | None = None) -> dict[str, ProcessRecord]:
    raw = _load_json_map(_path(root, "processes.yaml"))
    return {alias: ProcessRecord.from_dict(value) for alias, value in raw.items()}


def save_processes(records: dict[str, ProcessRecord], root: Path | None = None) -> None:
    _save_json_map(
        _path(root, "processes.yaml"), {k: records[k].to_dict() for k in sorted(records)}
    )


def load_admin_profiles(root: Path | None = None) -> dict[str, AdminProfile]:
    raw = _load_json_map(admin_profiles_path(root))
    profiles = raw.get("profiles", raw)
    if not isinstance(profiles, dict):
        raise StateError(f"state file {admin_profiles_path(root)} must contain a profiles object")
    return {
        name: AdminProfile.from_dict(value)
        for name, value in profiles.items()
        if isinstance(value, dict)
    }


def save_admin_profiles(
    records: dict[str, AdminProfile],
    root: Path | None = None,
    *,
    active_profile: str = "",
) -> None:
    _save_json_map(
        admin_profiles_path(root),
        {
            "active_profile": active_profile,
            "profiles": {k: records[k].to_dict() for k in sorted(records)},
        },
    )


def append_history(
    action: str, alias: str, tunnel_id: str | None, detail: str = "", root: Path | None = None
) -> None:
    path = _path(root, "history.md")
    path.parent.mkdir(parents=True, exist_ok=True)
    clean_detail = detail.replace("\n", " ").strip()
    line = f"- {utc_now()} action={action} alias={alias} tunnel_id={tunnel_id or '-'}"
    if clean_detail:
        line = f"{line} detail={clean_detail}"
    with path.open("a", encoding="utf-8") as f:
        f.write(line + "\n")


def reject_inline_secret_material(value: str, *, field: str) -> None:
    for pattern in _SECRET_PATTERNS:
        if pattern.search(value):
            raise StateError(
                f"{field} appears to contain inline secret material; use env or file references instead"
            )
    for token, next_token in _command_token_pairs(value):
        if not _SECRET_FLAG_PATTERN.match(token):
            continue
        if "=" in token:
            continue
        if next_token and not next_token.startswith("-") and not _is_secret_reference(next_token):
            raise StateError(
                f"{field} appears to contain inline secret material; use env or file references instead"
            )


def _command_token_pairs(value: str) -> list[tuple[str, str]]:
    try:
        tokens = shlex.split(value)
    except ValueError:
        tokens = value.split()
    return [
        (token, tokens[index + 1] if index + 1 < len(tokens) else "")
        for index, token in enumerate(tokens)
    ]


def _is_secret_reference(value: str) -> bool:
    return bool(_SECRET_REFERENCE_PATTERN.match(value))


def validate_secret_reference(value: str, *, field: str) -> None:
    if value.startswith(("env:", "file:")):
        return
    raise StateError(f"{field} must be an env:NAME or file:/path reference")


def alias_record_from_tunnel(
    *,
    alias: str,
    tunnel: dict[str, Any],
    admin_profile: str = "",
    description: str = "",
    config_path: str = "",
    health_url_file: str = "",
) -> AliasRecord:
    return AliasRecord(
        alias=alias,
        tunnel_id=str(tunnel.get("id", "")),
        name=str(tunnel.get("name", alias)),
        admin_profile=admin_profile,
        description=str(tunnel.get("description", description)),
        organization_ids=tuple(str(v) for v in tunnel.get("organization_ids", [])),
        workspace_ids=tuple(str(v) for v in tunnel.get("workspace_ids", [])),
        tenant_ids=tuple(str(v) for v in tunnel.get("tenant_ids", [])),
        config_path=config_path,
        health_url_file=health_url_file,
        updated_at=utc_now(),
    )


def _path(root: Path | None, name: str) -> Path:
    return ensure_state_dirs(root) / name


def _load_json_map(path: Path) -> dict[str, Any]:
    if not path.exists():
        return {}
    try:
        data = json.loads(path.read_text(encoding="utf-8"))
    except json.JSONDecodeError as exc:
        raise StateError(
            f"state file {path} is not valid JSON-compatible YAML; repair or remove it"
        ) from exc
    if not isinstance(data, dict):
        raise StateError(f"state file {path} must contain an object")
    return data


def _save_json_map(path: Path, data: dict[str, Any]) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    tmp = path.with_suffix(path.suffix + ".tmp")
    tmp.write_text(json.dumps(data, indent=2, sort_keys=True) + "\n", encoding="utf-8")
    os.chmod(tmp, 0o600)
    tmp.replace(path)

File: pkg/codexplugin/state/state.go
package state

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"runtime"
	"sort"
	"strings"
	"time"
)

var (
	aliasCharsPattern = regexp.MustCompile(`[^a-z0-9-]+`)
	secretPatterns    = []*regexp.Regexp{
		regexp.MustCompile(`sk-[A-Za-z0-9_-]{12,}`),
		regexp.MustCompile(`(?i)\bBearer\s+[A-Za-z0-9._~+/=-]{12,}`),
		regexp.MustCompile(`(?i)\bAuthorization\s*:`),
		regexp.MustCompile(`(?i)(api[_-]?key|access[_-]?token|refresh[_-]?token|id[_-]?token|password|secret)=\S+`),
		regexp.MustCompile(`(?i)\b(OPENAI_ADMIN_KEY|OPENAI_API_KEY|CONTROL_PLANE_API_KEY)=`),
		regexp.MustCompile(`://[^/\s:@]+:[^/\s:@]+@`),
	}
	secretFlagPattern      = regexp.MustCompile(`(?i)^-{1,2}[a-z0-9_-]*(api[_-]?key|access[_-]?token|refresh[_-]?token|id[_-]?token|password|secret)[a-z0-9_-]*$`)
	secretReferencePattern = regexp.MustCompile(`^(env:[A-Za-z_][A-Za-z0-9_]*|file:.+|\$[A-Za-z_][A-Za-z0-9_]*|\$\{[A-Za-z_][A-Za-z0-9_]*\})$`)
)

type Error struct {
	message string
}

func (e *Error) Error() string {
	return e.message
}

func NewError(message string) error {
	return &Error{message: message}
}

type AliasRecord struct {
	Alias           string   `json:"alias"`
	TunnelID        string   `json:"tunnel_id"`
	Name            string   `json:"name"`
	AdminProfile    string   `json:"admin_profile,omitempty"`
	Description     string   `json:"description,omitempty"`
	OrganizationIDs []string `json:"organization_ids,omitempty"`
	WorkspaceIDs    []string `json:"workspace_ids,omitempty"`
	TenantIDs       []string `json:"tenant_ids,omitempty"`
	ConfigPath      string   `json:"config_path,omitempty"`
	ProfileName     string   `json:"profile_name,omitempty"`
	ProfileDir      string   `json:"profile_dir,omitempty"`
	ProfilePath     string   `json:"profile_path,omitempty"`
	HealthURLFile   string   `json:"health_url_file,omitempty"`
	UpdatedAt       string   `json:"updated_at,omitempty"`
}

type ProcessRecord struct {
	Alias         string `json:"alias"`
	TunnelID      string `json:"tunnel_id"`
	ConfigPath    string `json:"config_path"`
	HealthURLFile string `json:"health_url_file"`
	TargetKind    string `json:"target_kind"`
	TargetValue   string `json:"target_value"`
	Command       string `json:"command"`
	StartedAt     string `json:"started_at"`
	AdminProfile  string `json:"admin_profile,omitempty"`
	ProfileName   string `json:"profile_name,omitempty"`
	ProfileDir    string `json:"profile_dir,omitempty"`
	ProfilePath   string `json:"profile_path,omitempty"`
	Mode          string `json:"mode,omitempty"`
	SessionName   string `json:"session_name,omitempty"`
	PID           int    `json:"pid,omitempty"`
	LogPath       string `json:"log_path,omitempty"`
}

type AdminProfile struct {
	Name                string `json:"name"`
	ControlPlaneBaseURL string `json:"control_plane_base_url"`
	ControlPlaneURLPath string `json:"control_plane_url_path,omitempty"`
	AdminKey            string `json:"admin_key"`
	UpdatedAt           string `json:"updated_at,omitempty"`
}

type AdminProfilesFile struct {
	ActiveProfile string                  `json:"active_profile,omitempty"`
	Profiles      map[string]AdminProfile `json:"profiles"`
}

type Root struct {
	Path string
}

func UTCNow() string {
	return time.Now().UTC().Truncate(time.Second).Format(time.RFC3339)
}

func NormalizeAlias(alias string) (string, error) {
	lowered := strings.ToLower(strings.TrimSpace(alias))
	lowered = strings.ReplaceAll(lowered, "_", "-")
	lowered = aliasCharsPattern.ReplaceAllString(lowered, "-")
	lowered = regexp.MustCompile(`-{2,}`).ReplaceAllString(lowered, "-")
	lowered = strings.Trim(lowered, "-")
	if lowered == "" {
		return "", NewError("alias must contain at least one ASCII letter or number")
	}
	return lowered, nil
}

func ResolveRoot(lookupEnv func(string) (string, bool)) Root {
	if lookupEnv == nil {
		lookupEnv = os.LookupEnv
	}
	if value, ok := lookupEnv("TUNNEL_CLIENT_STATE_DIR"); ok && strings.TrimSpace(value) != "" {
		return Root{Path: filepath.Clean(strings.TrimSpace(value))}
	}
	for _, candidate := range legacyRootCandidates(lookupEnv) {
		if rootExists(candidate) {
			return Root{Path: candidate}
		}
	}
	if value, ok := lookupEnv("XDG_STATE_HOME"); ok && strings.TrimSpace(value) != "" {
		return Root{Path: filepath.Join(filepath.Clean(strings.TrimSpace(value)), "tunnel-client")}
	}
	if value, ok := lookupEnv("HOME"); ok && strings.TrimSpace(value) != "" {
		return Root{Path: defaultStatePath(filepath.Clean(strings.TrimSpace(value)))}
	}
	if home, err := os.UserHomeDir(); err == nil && strings.TrimSpace(home) != "" {
		return Root{Path: defaultStatePath(filepath.Clean(home))}
	}
	return Root{Path: filepath.Join(".", ".tunnel-client")}
}

func legacyRootCandidates(lookupEnv func(string) (string, bool)) []string {
	candidates := []string{}
	if value, ok := lookupEnv("CODEX_HOME"); ok && strings.TrimSpace(value) != "" {
		candidates = append(candidates, filepath.Join(filepath.Clean(strings.TrimSpace(value)), "tunnel-mcp"))
	}
	if value, ok := lookupEnv("HOME"); ok && strings.TrimSpace(value) != "" {
		home := filepath.Clean(strings.TrimSpace(value))
		candidates = append(candidates,
			filepath.Join(home, ".codex", "tunnel-mcp"),
			filepath.Join(home, ".tunnel-client"),
		)
	}
	return candidates
}

func defaultStatePath(home string) string {
	if runtime.GOOS == "darwin" {
		return filepath.Join(home, "Library", "Application Support", "tunnel-client")
	}
	return filepath.Join(home, ".local", "state", "tunnel-client")
}

func rootExists(path string) bool {
	info, err := os.Stat(path)
	return err == nil && info.IsDir()
}

func EnsureDirs(root Root) error {
	for _, subdir := range []string{
		root.Path,
		filepath.Join(root.Path, "configs"),
		filepath.Join(root.Path, "health"),
		filepath.Join(root.Path, "logs"),
	} {
		if err := os.MkdirAll(subdir, 0o755); err != nil {
			return fmt.Errorf("create state directory %s: %w", subdir, err)
		}
	}
	return nil
}

func AliasesPath(root Root) string {
	return filepath.Join(root.Path, "aliases.yaml")
}

func ProcessesPath(root Root) string {
	return filepath.Join(root.Path, "processes.yaml")
}

func AdminProfilesPath(root Root) string {
	return filepath.Join(root.Path, "admin_profiles.yaml")
}

func HistoryPath(root Root) string {
	return filepath.Join(root.Path, "history.md")
}

func LoadAliases(root Root) (map[string]AliasRecord, error) {
	raw := map[string]AliasRecord{}
	if err := loadJSONMap(AliasesPath(root), &raw); err != nil {
		return nil, err
	}
	return raw, nil
}

func SaveAliases(root Root, records map[string]AliasRecord) error {
	sorted := make(map[string]AliasRecord, len(records))
	keys := make([]string, 0, len(records))
	for key := range records {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		sorted[key] = records[key]
	}
	return saveJSONMap(AliasesPath(root), sorted)
}

func LoadProcesses(root Root) (map[string]ProcessRecord, error) {
	raw := map[string]ProcessRecord{}
	if err := loadJSONMap(ProcessesPath(root), &raw); err != nil {
		return nil, err
	}
	return raw, nil
}

func SaveProcesses(root Root, records map[string]ProcessRecord) error {
	sorted := make(map[string]ProcessRecord, len(records))
	keys := make([]string, 0, len(records))
	for key := range records {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		sorted[key] = records[key]
	}
	return saveJSONMap(ProcessesPath(root), sorted)
}

func LoadAdminProfiles(root Root) (AdminProfilesFile, error) {
	if _, err := os.Stat(AdminProfilesPath(root)); err != nil {
		if os.IsNotExist(err) {
			return AdminProfilesFile{Profiles: map[string]AdminProfile{}}, nil
		}
		return AdminProfilesFile{}, fmt.Errorf("read state file %s: %w", AdminProfilesPath(root), err)
	}
	var raw map[string]json.RawMessage
	if err := loadJSONMap(AdminProfilesPath(root), &raw); err != nil {
		return AdminProfilesFile{}, err
	}
	file := AdminProfilesFile{Profiles: map[string]AdminProfile{}}
	if active, ok := raw["active_profile"]; ok {
		if err := json.Unmarshal(active, &file.ActiveProfile); err != nil {
			return AdminProfilesFile{}, &Error{message: fmt.Sprintf("state file %s must contain a string active_profile", AdminProfilesPath(root))}
		}
	}
	if profilesRaw, ok := raw["profiles"]; ok {
		if err := json.Unmarshal(profilesRaw, &file.Profiles); err != nil {
			return AdminProfilesFile{}, &Error{message: fmt.Sprintf("state file %s must contain a profiles object", AdminProfilesPath(root))}
		}
		return file, nil
	}
	for key, value := range raw {
		if key == "active_profile" {
			continue
		}
		var profile AdminProfile
		if err := json.Unmarshal(value, &profile); err != nil {
			return AdminProfilesFile{}, &Error{message: fmt.Sprintf("state file %s must contain a profiles object", AdminProfilesPath(root))}
		}
		file.Profiles[key] = profile
	}
	return file, nil
}

func SaveAdminProfiles(root Root, file AdminProfilesFile) error {
	if file.Profiles == nil {
		file.Profiles = map[string]AdminProfile{}
	}
	sorted := make(map[string]AdminProfile, len(file.Profiles))
	keys := make([]string, 0, len(file.Profiles))
	for key := range file.Profiles {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	for _, key := range keys {
		sorted[key] = file.Profiles[key]
	}
	return saveJSONMap(AdminProfilesPath(root), map[string]any{
		"active_profile": file.ActiveProfile,
		"profiles":       sorted,
	})
}

func AppendHistory(root Root, action, alias, tunnelID, detail string) error {
	if err := EnsureDirs(root); err != nil {
		return err
	}
	line := fmt.Sprintf("- %s action=%s alias=%s tunnel_id=%s", UTCNow(), action, alias, valueOrDash(tunnelID))
	if clean := strings.TrimSpace(strings.ReplaceAll(detail, "\n", " ")); clean != "" {
		line = line + " detail=" + clean
	}
	f, err := os.OpenFile(HistoryPath(root), os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0o600)
	if err != nil {
		return fmt.Errorf("append history %s: %w", HistoryPath(root), err)
	}
	if _, err := f.WriteString(line + "\n"); err != nil {
		_ = f.Close()
		return fmt.Errorf("append history %s: %w", HistoryPath(root), err)
	}
	if err := f.Close(); err != nil {
		return fmt.Errorf("close history %s: %w", HistoryPath(root), err)
	}
	return nil
}

func RejectInlineSecretMaterial(value string, field string) error {
	for _, pattern := range secretPatterns {
		if pattern.MatchString(value) {
			return &Error{message: fmt.Sprintf("%s appears to contain inline secret material; use env or file references instead", field)}
		}
	}
	for _, pair := range commandTokenPairs(value) {
		token := pair[0]
		next := pair[1]
		if !secretFlagPattern.MatchString(token) || strings.Contains(token, "=") {
			continue
		}
		if next != "" && !strings.HasPrefix(next, "-") && !IsSecretReference(next) {
			return &Error{message: fmt.Sprintf("%s appears to contain inline secret material; use env or file references instead", field)}
		}
	}
	return nil
}

func ValidateSecretReference(value string, field string) error {
	if strings.HasPrefix(value, "env:") || strings.HasPrefix(value, "file:") {
		return nil
	}
	return NewError(fmt.Sprintf("%s must be an env:NAME or file:/path reference", field))
}

func IsSecretReference(value string) bool {
	return secretReferencePattern.MatchString(value)
}

func AliasRecordFromTunnel(
	alias string,
	tunnelID string,
	name string,
	description string,
	organizationIDs []string,
	workspaceIDs []string,
	tenantIDs []string,
	adminProfile string,
	configPath string,
	profileName string,
	profileDir string,
	profilePath string,
	healthURLFile string,
) AliasRecord {
	return AliasRecord{
		Alias:           alias,
		TunnelID:        tunnelID,
		Name:            name,
		AdminProfile:    adminProfile,
		Description:     description,
		OrganizationIDs: append([]string(nil), organizationIDs...),
		WorkspaceIDs:    append([]string(nil), workspaceIDs...),
		TenantIDs:       append([]string(nil), tenantIDs...),
		ConfigPath:      configPath,
		ProfileName:     profileName,
		ProfileDir:      profileDir,
		ProfilePath:     profilePath,
		HealthURLFile:   healthURLFile,
		UpdatedAt:       UTCNow(),
	}
}

func loadJSONMap(path string, dest any) error {
	data, err := os.ReadFile(path)
	if err != nil {
		if os.IsNotExist(err) {
			return nil
		}
		return fmt.Errorf("read state file %s: %w", path, err)
	}
	if err := json.Unmarshal(data, dest); err != nil {
		return &Error{message: fmt.Sprintf("state file %s is not valid JSON-compatible YAML; repair or remove it", path)}
	}
	return nil
}

func saveJSONMap(path string, data any) error {
	if err := os.MkdirAll(filepath.Dir(path), 0o755); err != nil {
		return fmt.Errorf("create state directory for %s: %w", path, err)
	}
	payload, err := json.MarshalIndent(data, "", "  ")
	if err != nil {
		return fmt.Errorf("marshal state file %s: %w", path, err)
	}
	tmp := path + ".tmp"
	if err := os.WriteFile(tmp, append(payload, '\n'), 0o600); err != nil {
		return fmt.Errorf("write state file %s: %w", tmp, err)
	}
	if err := os.Rename(tmp, path); err != nil {
		return fmt.Errorf("replace state file %s: %w", path, err)
	}
	return nil
}

func commandTokenPairs(value string) [][2]string {
	tokens := strings.Fields(value)
	pairs := make([][2]string, 0, len(tokens))
	for i, token := range tokens {
		next := ""
		if i+1 < len(tokens) {
			next = tokens[i+1]
		}
		pairs = append(pairs, [2]string{token, next})
	}
	return pairs
}

func valueOrDash(value string) string {
	if strings.TrimSpace(value) == "" {
		return "-"
	}
	return value
}
